In [58]:
import numpy as np
import pandas as pd
from matplotlib.pyplot import subplots
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor as VIF
from statsmodels.stats.anova import anova_lm

In [59]:
from ISLP import load_data
from ISLP.models import (ModelSpec as MS, summarize, poly)

ISLP Applied exercise
Q8: Using the Auto data set. 

In [60]:
Auto=load_data('Auto')
Auto

,mpg,cylinders,displacement,horsepower,weight,acceleration,year,origin
name,,,,,,,,
chevrolet chevelle malibu,18.0,8,307.0,130,3504,12.0,70,1
buick skylark 320,15.0,8,350.0,165,3693,11.5,70,1
plymouth satellite,18.0,8,318.0,150,3436,11.0,70,1
amc rebel sst,16.0,8,304.0,150,3433,12.0,70,1
ford torino,17.0,8,302.0,140,3449,10.5,70,1
...,...,...,...,...,...,...,...,...
ford mustang gl,27.0,4,140.0,86,2790,15.6,82,1
vw pickup,44.0,4,97.0,52,2130,24.6,82,2
dodge rampage,32.0,4,135.0,84,2295,11.6,82,1


In [61]:
# Simple linear regression using mpg as response and horsepower as the predictor. 
# setting up an additional column that consists of ones which represent an intercept constant. 
X=pd.DataFrame({'intercept':np.ones(Auto.shape[0]), 'horsepower':Auto['horsepower']})
X[:6]


,intercept,horsepower
name,,
chevrolet chevelle malibu,1.0,130
buick skylark 320,1.0,165
plymouth satellite,1.0,150
amc rebel sst,1.0,150
ford torino,1.0,140
ford galaxie 500,1.0,198


In [62]:
# setting up the response variable
y=Auto['mpg']

In [63]:
model=sm.OLS(y, X)
result=model.fit()
summarize(result)


,coef,std err,t,P>|t|
intercept,39.9359,0.717,55.660,0.0
horsepower,-0.1578,0.006,-24.489,0.0


In [64]:
design=MS(['horsepower'])
X=design.fit_transform(Auto)
X[:4]

,intercept,horsepower
name,,
chevrolet chevelle malibu,1.0,130
buick skylark 320,1.0,165
plymouth satellite,1.0,150
amc rebel sst,1.0,150


In [65]:
new_df=pd.DataFrame({'horsepower':[98]})

# the model has already learnt the rules and calculated statistical metrics like std dev on the entire dataset, therefore we only call transform and not fit and transform
newX=design.transform(new_df)
newX

,intercept,horsepower
0,1.0,98


In [71]:
new_predictions=result.get_prediction(newX);
prediction_summary=new_predictions.summary_frame(alpha=0.05)
print(prediction_summary)

        mean   mean_se  mean_ci_lower  mean_ci_upper  obs_ci_lower  \
0  24.467077  0.251262      23.973079      24.961075     14.809396   

   obs_ci_upper  
0     34.124758  


i. Yes, p-value is 0.00 (<0.05), indicating a statistically significant relationship.

ii. Very strong relationship, indicated by a massive t-statistic of -24.489.

iii. Negative relationship (coefficient = -0.1578); higher horsepower predicts lower mpg.

iv. CI=[LL:23.973079, UL:24.961075] Prediction interval=[lower: 14.809396, upper:34.124758  ]
